include/onnxruntime/core/common/make_string.h

```diff
@@ -36,6 +38,39 @@ inline void MakeStringImpl(std::ostringstream& ss, const T& t, const Args&... ar
   MakeStringImpl(ss, t);
   MakeStringImpl(ss, args...);
 }
+
+// see MakeString comments for explanation of why this is necessary
+template <typename... Args>
+inline std::string MakeStringImpl(const Args&... args) noexcept {
+  std::ostringstream ss;
+  MakeStringImpl(ss, args...);
+  return ss.str();
+}
+
+//
+// Infrastructure to convert char[n] to char* to reduce binary size
+//
+
+// default is to leave the type as is
+template <class T>
+struct if_char_array_make_ptr {
+  using type = T;
+};
+
+// specialization that matches an array reference, which is what the char array from a string literal 
+// used in a call to MakeString will be.
+// if the type is a char[n] array we 'decay' it to a char* so that the usages can be folded.
+template <class T, size_t N>
+struct if_char_array_make_ptr<T (&)[N]> {
+  // remove a single extent (T[x] -> T, but T[x][y] -> T[y]) so we only match char[x],
+  // and get the type name without the 'const' so both 'const char (&)[n]' and 'char (&)[n]' are matched. 
+  using element_type = std::remove_const<std::remove_extent<T>::type>::type;
+  using type = std::conditional<std::is_same<char, element_type>::value, const T*, T (&)[N]>::type;
```

Comment:
```suggestion
  using element_type = typename std::remove_const<typename std::remove_extent<T>::type>::type;
  using type = typename std::conditional<std::is_same<char, element_type>::value, const T*, T (&)[N]>::type;
```

Reply:
Is this needed for a specific build? I don't recall seeing the usage of 'typename' in conjunction with the '::type' value from these helpers.

Reply:
Weird. I see the Android CI that builds on MacOS wants it. Cross compiling Android on Windows was perfectly happy without that.

Reply:
Without this the build on my mac failed, sees Windows CI run failed for the same reason